Feat/add user auth

abungay opened the pull request:
## Issues
#26, #32 
## Description of changes
Add initial backend support for user authentication. Currently supporting the normal Register (admin doesn't yet generate a special token for admin registration), login, and logout features. Blocks the user from accessing any pages except login, register etc. until they are logged in.

Validates User accounts before they are saved during the register process:
* firstName and lastName must not be empty
* email must be a valid email address
* email must not already be in use
* password must be between 8-72 characters
* password and password confirmation must match on registration page

On login page:
* email and password must be valid

Security features:
* Other pages cannot be accessed until the user is logged in.
* CSRF (Cross-Site Request Forgery) protection from all login/logout/register forms
* Passwords are hashed with a salt using the BCrypt hashing algorithm

Added resources/data.sql to insert the initial roles into the role table of the database upon first initialization.
Added validation.properties file to store common validation messages (currently in use by validator/UserValidator.java).

Admin registration and forgot-password features will be added in a future PR since they require tokens and emails to be sent.
## Screenshots (If Applicable)
![register_validation](https://user-images.githubusercontent.com/5737382/55130086-f50a2780-50ef-11e9-901e-95e36ab49521.PNG)
![login_validation](https://user-images.githubusercontent.com/5737382/55130091-f8051800-50ef-11e9-8b7b-cfbc1b34560d.PNG)


## Checklist
- [x] I have unit tested my code.
- [x] I have integration tested my code.
- [x] All new and existing tests passed.
- [x] I have [updated the diagrams](https://github.com/amazin-team/Amazin-online-bookstore/wiki/Updating-the-diagrams) ModelsUMLClassDiagram and EntityRelationshipDiagram (if necessary).

abungay committed:
wip: add initial files for user-auth

abungay committed:
fix: get rid of bad annotation errors, update login form

abungay committed:
fix: remove unused import

LauraMachado21 commented on src/main/java/amazin/controller/UserController.java:
```diff
@@ -0,0 +1,74 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(UserController.class);
+
+    @GetMapping("/login")
```
Good refactoring!

abungay committed:
fix: ensure models and validation being added for each user controller

abungay committed:
fix: ensure models and validation being added for each user controller

abungay committed:
fix: properly assign model fields to thymeleaf form

LauraMachado21 commented on src/main/java/amazin/model/Role.java:
```diff
@@ -0,0 +1,39 @@
+    private Long id;
+    private String name;
+    @ManyToMany(mappedBy = "roles")
+    private Set<User> users;
```
How come you put the users in the Role? Rather than the role type in the User?

abungay replied:
The role has a name and an id and the mapped by I'm just saying that will come from the users.

role type is also in user its just a many-many relationship so both classes have a ref to the other one

LauraMachado21 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,76 @@
+    private String password;
+    private String passwordConfirmation;
+    @ManyToMany
+    private Set<Role> roles;
```
In which case will Users have more than 1 role?

abungay replied:
More so for future things but they will have user and admin roles at a maximum for now

Hajar1 replied:
I'd concur @LauraMachado21 statement. 
Users privilege hierarchy should be structured in a ring style hierarchy. 
Here's some food for thoughts about the subject: https://en.wikipedia.org/wiki/Protection_ring

abungay committed:
Merge remote-tracking branch 'remotes/origin/master' into feat/add-user-auth

# Conflicts:
#	pom.xml
#	src/main/java/amazin/model/Book.java

abungay committed:
rename column names similar to Book model

abungay committed:
fix: rename tables so they don't conflict with postgres reserved words

abungay committed:
add unit tests for User class

abungay committed:
fix typo in User test email case

abungay committed:
add Role class unit tests

abungay committed:
update diagrams for new users and roles tables

abungay commented:
~~Currently having issues with postgres saying that user0.id is not a valid column.~~

abungay committed:
fix: remove @NotNull validation from roles field

This was causing the registration form to not work with PostgreSQL set up since the form would submit roles as null.

zaidoon1 commented on src/main/java/amazin/model/Book.java:
```diff
@@ -12,7 +12,7 @@
 @Entity
 @Indexed
-@Table(name="book")
+@Table(name="book_table")
```
Renaming the table is not necessary. Table names don't typically have a _table prefix.

abungay replied:
I was following convention here since I think I had to rename the user table to user_table since user seems to be a reserved keyword in postgreSQL. I figured instead of only having the user table with the _table prefix, I would make all of them have the _table prefix.

zaidoon1 commented on src/main/java/amazin/model/Role.java:
```diff
@@ -0,0 +1,49 @@
+
+@Entity
+@Indexed
+@Table(name="role_table")
```
Table names don't typically have a _table prefix.

zaidoon1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,96 @@
+
+@Entity
+@Indexed
+@Table(name="user_table")
```
Table names don't typically have a _table prefix.

zaidoon1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,96 @@
+    @Column(name = "user_id", updatable = false, nullable = false)
+    private Long id;
+    @NotNull
+    @Column(name="user_first_name")
```
you are defining this in the user table, no need to say user_first_name, first_name is sufficient.

abungay replied:
following the convention here as used in Book.java -> e.g. book_id, book_name, etc.

zaidoon1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,96 @@
+    @Field
+    private String firstName;
+    @NotNull
+    @Column(name="user_last_name")
```
you are defining this in the user table, no need to say user_last_name, last_name is sufficient.

zaidoon1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,96 @@
+    private String lastName;
+    @NotNull
+    @Field
+    @Column(unique = true, name="user_email")
```
you are defining this in the user table, no need to say user_email, email is sufficient.

zaidoon1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,96 @@
+    @Column(unique = true, name="user_email")
+    private String email;
+    @NotNull
+    @Column(name="user_password")
```
you are defining this in the user table, no need to say user_password, password is sufficient.

zaidoon1 reviewed (commented):
In that case, it's good. Thanks!

abungay committed:
refactor: move WebSecurityConfig to security package

abungay committed:
fix: name field in Role should be unique

abungay committed:
feat: add AuthoritiesConstants for User Roles

abungay committed:
refactor: remove id column from Role table, only need role_name

abungay committed:
feat: add data.sql file to init DB Role table with initial roles

abungay committed:
feat: add findByName() method to RoleRepository

abungay committed:
fix: only assign USER role upon saving User to DB

ADMIN role will be conditionally set in a later PR

abungay committed:
feat: validate User fields

> ensure firstname, lastname not empty
> ensure username follows username format
> ensure username not taken
> ensure password between 8 and 72 characters
> ensure passwords match

abungay committed:
feat: display error messages to users on registration page next to fields

abungay committed:
feat: add more authentication/security tests for services

abungay committed:
fix: add @Test annotation above createTest for BookServiceTest

abungay committed:
refactor: remove code from /admin/register endpoint for now, not working yet

abungay committed:
feat: add tests for UserController get endpoints

abungay committed:
feat: add UserController tests for Post endpoints

-> added spring-security-test to pom.xml since we need it to test endpoints protected by CSRF

abungay committed:
feat: update diagrams

abungay commented:
note: since this introduces schema changes to the database we will need to drop the existing tables upon the first deploy to Heroku then everything should work as expected.

zaidoon1 reviewed (dismissed):
looks good! Nice work

MohamedZalat commented on src/main/java/amazin/controller/UserController.java:
```diff
@@ -0,0 +1,86 @@
+    public String register(@Valid @ModelAttribute User userForm, BindingResult result, Model model) {
+        userValidator.validate(userForm, result);
+        if (result.hasErrors()) {
+            return "registration";
```
This needs to be a constant variable as you used it twice.

MohamedZalat commented on src/main/java/amazin/controller/UserController.java:
```diff
@@ -0,0 +1,86 @@
+        userService.save(userForm);
+        securityService.autoLogin(userForm.getEmail(), userForm.getPasswordConfirmation());
+
+        return "redirect:/";
```
This also needs a constant.

MohamedZalat commented on src/main/java/amazin/model/Book.java:
```diff
@@ -12,47 +12,37 @@
 
 @Entity
 @Indexed
-@Table(name="book")
```
Why did you remove the Table and Column annotations? They are required for postgresql.

abungay replied:
they are not required for postgresql and add clutter to the file, it automatically stores the table Book as book columns go from camelCase to snake_case automatically, the only table I need to rename is the User table since user is a reserved keyword for postgresql

MohamedZalat commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long id;
```
Column annotation need to be used here in every field you need to store. Give each column an explicit id so we can do efficient joins when book recommendation needs to be implemented.

abungay replied:
you can still perform joins even if the columns are not renamed

MohamedZalat commented on src/main/java/amazin/model/Role.java:
```diff
@@ -0,0 +1,34 @@
+
+@Entity
+@Indexed
+public class Role {
```
Needs an @Table annotation and @Column annotation in fields you need to store too. You made a repository interface for it yet didn't annotate it as a table.

abungay replied:
@Table and @Column are not required

MohamedZalat replied:
It is important to annotate the Role with Table so we know it is a stored table.

abungay replied:
so we know it is stored? won't all the entities be stored?

MohamedZalat commented on src/main/java/amazin/security/WebSecurityConfig.java:
```diff
@@ -0,0 +1,50 @@
+    protected void configure(HttpSecurity http) throws Exception {
+        http
+            .authorizeRequests()
+                .antMatchers("/webjars/**", "/register", "/admin/register", "/forgot-password").permitAll()
```
What is the webjars endpoint? Why is it there?

abungay replied:
these are the static resources eg. CSS files that need to be accessible to the user before they login so that the login and register pages have proper css syling

MohamedZalat replied:
Sounds good.

Hajar1 replied:
Really neat if you thought about this. Thumbsup

MohamedZalat commented on src/main/java/amazin/security/WebSecurityConfig.java:
```diff
@@ -0,0 +1,50 @@
+        http
+            .authorizeRequests()
+                .antMatchers("/webjars/**", "/register", "/admin/register", "/forgot-password").permitAll()
+                .anyRequest().authenticated()
```
The index also needs you to be authenticated? Why?

abungay replied:
are you saying we should allow people to be allowed to visit the homepage without logging in as anonymous users? If so this is a different issue and we can tackle this in a separate pr

MohamedZalat replied:
Yes they need to visit the homepage so they click the log in link. It is not a different issue. Not every user that uses our website need to type /login instead of clicking a login button on the homepage with limited access.

abungay replied:
The default behaviour of spring security is for the user to have to log in before doing anything else on the webpage. When the user is not logged in any url they try to access will redirect them automatically to the login page. For the feature of letting them also see the books and everything before logging in this could be done but I suggest this is done in a future PR

MohamedZalat commented on src/main/java/amazin/service/SecurityServiceImpl.java:
```diff
@@ -0,0 +1,45 @@
+    @Override
+    public void autoLogin(String username, String password) {
+        UserDetails userDetails = userDetailsService.loadUserByUsername(username);
+        UsernamePasswordAuthenticationToken usernamePasswordAuthenticationToken = new UsernamePasswordAuthenticationToken(userDetails, password, userDetails.getAuthorities());
```
This line is way too long. Please break it into 2 lines.

MohamedZalat committed:
format(SecurityServiceImpl): Break long line of code

abungay committed:
fix: don't add _table suffix to table names for jpa

abungay committed:
update data.sql DB initializer for updated role table name

abungay committed:
Merge branch 'master' of https://github.com/amazin-team/Amazin-online-bookstore into feat/add-user-auth

Hajar1 commented on src/main/java/amazin/controller/ApplicationController.java:
```diff
@@ -13,30 +13,9 @@
     @Autowired
     BookRepository bookRepository;
 
-    @GetMapping({ "/" })
```
unnecessary change

abungay replied:
I was refactoring this file anyways and the ones in the new Controller are using this format we don't need curly braces here

Hajar1 commented on src/main/java/amazin/controller/UserController.java:
```diff
@@ -0,0 +1,86 @@
+    @Autowired
+    private UserValidator userValidator;
+
+    private static final Logger logger = LoggerFactory.getLogger(UserController.class);
```
Logger defined but never used.

Assert that is not needed

Hajar1 commented on src/main/java/amazin/controller/UserController.java:
```diff
@@ -0,0 +1,86 @@
+import amazin.service.SecurityService;
+import amazin.service.UserService;
+import amazin.validator.UserValidator;
+import org.slf4j.Logger;
```
logger might not be useful

Hajar1 commented on src/main/java/amazin/controller/UserController.java:
```diff
@@ -0,0 +1,86 @@
+import amazin.service.UserService;
+import amazin.validator.UserValidator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Same thing

Hajar1 commented on src/main/java/amazin/model/Role.java:
```diff
@@ -0,0 +1,36 @@
+@Table(name="role")
+public class Role {
+
+    @Id
```
The name field should not be the field that we are basing ID on. 
Please use something around the lines:
 @Id @GeneratedValue long id;

zaidoon1 replied:
this is fine as long as it is unique which in our case it is

abungay replied:
we don't need an int id since it will never be used, user_name is the primary key here

Hajar1 replied:
> this is fine as long as it is unique which in our case it is

> we don't need an int id since it will never be used, user_name is the primary key here

I see what you mean. A part of me thinks then table role does not add much if it just keeps a track record of who are the ADMINS and who are the USERS. In a perfect world, I believe this table could be just replaced with a column in the user table but for the time being that should be fine

zaidoon1 replied:
this wouldn't work because we are assuming a user can have multiple roles and as such we will have data duplication in the table

zaidoon1 replied:
> > this is fine as long as it is unique which in our case it is
> 
> > we don't need an int id since it will never be used, user_name is the primary key here
> 
> I see what you mean. A part of me thinks then table role does not add much if it just keeps a track record of who are the ADMINS and who are the USERS. In a perfect world, I believe this table could be just replaced with a column in the user table but for the time being that should be fine

this wouldn't work because we are assuming a user can have multiple roles and as such we will have data duplication in the table

Hajar1 replied:
Well I also mentioned that users shouldnt have multiple roles you cant be ADMIN and a USER and an ANNONYMOUS that might not make sense how can i be annonymous and an admin?

zaidoon1 replied:
> Well I also mentioned that users shouldnt have multiple roles you cant be ADMIN and a USER and an ANNONYMOUS that might not make sense how can i be annonymous and an admin?

Aaron did this because he was thinking we might add more roles in the future

MohamedZalat replied:
This way is more abstract and therefore preferred. Abstraction is preferable in places where it makes sense. An example for future roles: a seller may also be a buyer. The rules you mentioned are just rules to be applied to the mapping, as they are exceptions to the norm.

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+public class User {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
```
You don't need to specify the strategy

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(name = "user_id", updatable = false, nullable = false)
```
is this useful?
have you tried without it?
Id will never be null if it is auto updated --> hence u don't need to do the check

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+    private Long id;
+    @NotNull
+    @Size(min = 1, message = "First name cannot be empty")
+    @Column(name="user_first_name")
```
I'd still say that using @tablename_@fieldname style is an overkill tbh but that's up to you :)

abungay replied:
yes @MohamedZalat wants it like this @zaidoon1 and I wanted this prefix removed

MohamedZalat replied:
I can revert this all and keep consistent with just the id field being prefixed with the table name, if you all want this. I still have Aaron's commit to remove all tags and I can add another one for the tables and the ids.

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+    @NotNull
+    @Size(min = 1, message = "Last name cannot be empty")
+    @Column(name="user_last_name")
+    @Field
```
Not necessary if you are not changing the name of the field

zaidoon1 replied:
same as last comment

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+    @NotNull
+    @Size(min = 1, message = "First name cannot be empty")
+    @Column(name="user_first_name")
+    @Field
```
Not necessary

zaidoon1 replied:
@Field is necessary if he wants that field to be indexed and thus searchable. I take it we will allow for searching of users then?

Hajar1 replied:
> @field is necessary if he wants that field to be indexed and thus searchable. I take it we will allow for searching of users then?

You don't need to specify both @column and @field  @zaidoon1   
Quote from Spring data annotation - [link](https://www.baeldung.com/spring-data-annotations)

> With @Field, we can configure the name of a field we want to use when MongoDB persists the document. Note, that this annotation is the Mongo equivalent of @Column in JPA.

zaidoon1 replied:
Hibernate search requires the @field annotation
![Screenshot_20190329_214144](https://user-images.githubusercontent.com/43054535/55269591-3edc4480-528d-11e9-945e-be433839b40d.png)

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+    private String lastName;
+    @NotNull
+    @Pattern(regexp = "^\\w+@[a-zA-Z_]+?\\.[a-zA-Z]{2,3}$", message = "Not a valid email address")
+    @Field
```
not useful

zaidoon1 replied:
@field is necessary if he wants that field to be indexed and thus searchable. I take it we will allow for searching of users then?

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+    @Size(min = 1, message = "Last name cannot be empty")
+    @Column(name="user_last_name")
+    @Field
+    private String lastName;
```
Please leave a space between fields or inline annotations

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+    @Field
+    private String lastName;
+    @NotNull
+    @Pattern(regexp = "^\\w+@[a-zA-Z_]+?\\.[a-zA-Z]{2,3}$", message = "Not a valid email address")
```
Your regex does not work on school emails.
I would opt for a simpler one as well.
a simple @ check might be enough 
though realistically something like: "^\S+@\S+$" should do the job

abungay replied:
oh is the @cmail.carleton.ca failing when you test it?

Hajar1 replied:
Yes, you could use the regex I provided. It should work

MohamedZalat replied:
This one is important.

zaidoon1 replied:
so when it comes to regex expressions, there is always a trade off, whether it's slow, or missing cases. The best one that is currently available is:

``(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])``

This works for 99.99% of email addresses which is more than enough for our use case. The link can be found below:
http://emailregex.com/

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+    @Size(min = 8, message = "Password must be at least 8 characters")
+    @Size(max = 72, message = "Password must be less than 72 characters")
+    @Column(name="user_password")
+    @Field
```
Not useful

zaidoon1 replied:
@field is necessary if he wants that field to be indexed and thus searchable. I take it we will allow for searching of users then?

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+    @Column(name="user_password")
+    @Field
+    private String password;
+    @NotNull
```
space or inline

abungay committed:
fix: indentation for long line in SecurityServiceImpl

abungay committed:
refactor: add constants for comonly used views in UserController

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+        this.passwordConfirmation = passwordConfirmation;
+    }
+
+    public Set<Role> getRoles() {
```
Role not set

abungay replied:
not set where? when the user is  getting saved in the user service it sets the ROLE_USER as their role for now

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+        return roles;
+    }
+
+    public void setRoles(Set<Role> roles) {
```
Role not set

abungay replied:
same as above

Hajar1 commented on src/main/java/amazin/repository/RoleRepository.java:
```diff
@@ -0,0 +1,13 @@
+
+@RepositoryRestResource(collectionResourceRel = "role", path = "role.json")
+public interface RoleRepository extends CrudRepository<Role, Long> {
+    Set<Role> findByName(String name);
```
Role not set

abungay replied:
same as above

Hajar1 commented on src/main/java/amazin/repository/RoleRepository.java:
```diff
@@ -0,0 +1,13 @@
+@RepositoryRestResource(collectionResourceRel = "role", path = "role.json")
+public interface RoleRepository extends CrudRepository<Role, Long> {
+    Set<Role> findByName(String name);
+    Set<Role> findAll();
```
Role not set

abungay replied:
same as above

Hajar1 commented on src/main/java/amazin/security/AuthoritiesConstants.java:
```diff
@@ -0,0 +1,13 @@
+
+public final class AuthoritiesConstants {
+
+    public static final String ADMIN = "ROLE_ADMIN";
```
I still think we should just name the column table the same name of the attribute

abungay replied:
I don't understand

Hajar1 commented on src/main/java/amazin/security/AuthoritiesConstants.java:
```diff
@@ -0,0 +1,13 @@
+
+    public static final String ANONYMOUS = "ROLE_ANONYMOUS";
+
+    private AuthoritiesConstants() {
```
do you have to define an empty constructor doesn't class object already does that for you?

zaidoon1 replied:
this constuctor is private, which is not the default (public). Is this needed by some library?

abungay replied:
defined as private since this is just constants, other spring security examples I found online are doing the same thing

zaidoon1 replied:
then it is fine

Hajar1 replied:
> defined as private since this is just constants, other spring security examples I found online are doing the same thing

Is this class then acting as a global variable? is this best practice?

MohamedZalat replied:
Everything in Java has to be a class.

Hajar1 commented on src/main/java/amazin/service/SecurityServiceImpl.java:
```diff
@@ -0,0 +1,47 @@
+
+        if (usernamePasswordAuthenticationToken.isAuthenticated()) {
+            SecurityContextHolder.getContext().setAuthentication(usernamePasswordAuthenticationToken);
+            logger.debug(String.format("Auto login %s successfully!", username));
```
Should we push to production code a debug statement?

zaidoon1 replied:
that's fine, the whole point of using a logger like this is so that you can run the application in debug mode to get more verbose output. This is pretty standard.

abungay replied:
doesn't really effect performance and is kind of helpful for debugging if we need it

Hajar1 replied:
debugging a working component?
wouldn't we end up with a log of users successfully logging in which clutter from important logs?

zaidoon1 replied:
> debugging a working component?
> wouldn't we end up with a log of users successfully logging in which clutter from important logs?

This is why it's a debug message, it means we want to know everything that's happening

MohamedZalat replied:
It is an important debug message as it logs the user in case of an issue caused, we would know the user affected. The level is configured to debug meaning only when verbosity is changed to debug we will see it.

Hajar1 commented on src/main/java/amazin/service/UserDetailsServiceImpl.java:
```diff
@@ -0,0 +1,39 @@
+    @Transactional(readOnly = true)
+    public UserDetails loadUserByUsername(String username) {
+        User user = userRepository.findByEmail(username);
+        if (user == null) {
```
Object-oriented way of checking for null:
Objects.isNull(user) 

Might be a better fit since we are in an object-oriented context. :)

abungay replied:
I have never seen that Objects.isNull method being used before

zaidoon1 replied:
what aaron has is fine

Hajar1 replied:
> what aaron has is fine

Saying fine for everything is not a way to review :p

zaidoon1 replied:
> > what aaron has is fine
> 
> Saying fine for everything is not a way to review :p

there is nothing wrong with user == null, it's pretty standard.
![Screenshot_20190329_215619](https://user-images.githubusercontent.com/43054535/55269768-0b9ab500-528f-11e9-9ffd-25385ff6b488.png)

Hajar1 commented on src/main/java/amazin/service/UserDetailsServiceImpl.java:
```diff
@@ -0,0 +1,39 @@
+        }
+
+        Set<GrantedAuthority> grantedAuthorities = new HashSet<>();
+        for (Role role : user.getRoles()){
```
might need to be turned to a simple set if you decide to follow the ring architecture of privilege

abungay replied:
simple set?

Hajar1 commented on src/main/java/amazin/service/UserService.java:
```diff
@@ -0,0 +1,8 @@
+import amazin.model.User;
+
+public interface UserService {
+    void save(User user);
```
Shouldn't we return success if saved? and false if not

MohamedZalat replied:
We can create this as a new issue.

Hajar1 commented on src/main/java/amazin/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,33 @@
+    public void save(User user) {
+        user.setRoles(new HashSet<>(roleRepository.findByName(AuthoritiesConstants.USER)));
+        user.setPassword(bCryptPasswordEncoder.encode(user.getPassword()));
+        userRepository.save(user);
```
what if this fails? should we not throw an error?

Hajar1 commented on src/main/java/amazin/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public User findByEmail(String email) {
+        return userRepository.findByEmail(email);
```
What if a user is not found?

abungay replied:
you get null

Hajar1 commented on src/main/java/amazin/validator/UserValidator.java:
```diff
@@ -0,0 +1,33 @@
+    private UserService userService;
+
+    @Override
+    public boolean supports(Class<?> aClass) {
```
Not so straight forward what u are trying to do here;
Might need a comment explaining the function.

Also function is not used; do we have to have it ?

abungay replied:
this is a method required for the Validator interface, you can look up the javadocs for org.springframework.validation.Validator if you want to know more about Validators or this method

Hajar1 commented on src/main/java/amazin/validator/UserValidator.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public void validate(Object o, Errors errors) {
+        User user = (User) o;
```
This might fail; I'd believe we must wrap it in a try and catch

abungay replied:
I don't think that's necessary here we have the supports method for this purpose to make sure its a User.class

Hajar1 replied:
If we are that confident then why not specify it in the function parameter as user rather then casting it inside of it ?

MohamedZalat replied:
Look at Validator interface.

Hajar1 commented on src/main/resources/templates/layout.html:
```diff
@@ -159,8 +159,7 @@
     <!-- Logout Modal-->
-    <div class="modal fade" id="logoutModal" tabindex="-1" role="dialog" aria-labelledby="exampleModalLabel"
-        aria-hidden="true">
+    <div class="modal fade" id="logoutModal" tabindex="-1" role="dialog" aria-labelledby="exampleModalLabel" aria-hidden="true">
```
unnecessary edit

abungay replied:
that's fine

zaidoon1 replied:
no problem with it

Hajar1 commented on src/main/resources/templates/layout.html:
```diff
@@ -169,10 +168,17 @@ <h5 class="modal-title" id="exampleModalLabel">Ready to Leave?</h5>
                     <button class="btn btn-secondary" type="button" data-dismiss="modal">Cancel</button>
-                    <a class="btn btn-primary" href="login">Logout</a>
+                    <button class="btn btn-primary modal-logout"
+                            onclick="event.preventDefault();
```
Why are you preventing default operation of submitting the form and then trying to submit the form?
why not use a submit input?

Hajar1 commented on src/main/resources/templates/registration-admin.html:
```diff
@@ -1,5 +1,5 @@
 <!DOCTYPE html>
-<html lang="en">
+<html lang="en" xmlns:th="http://www.w3.org/1999/xhtml" xmlns:layout="http://www.ultraq.net.nz/thymeleaf/layout">
```
Why does this file only contain a head not a whole admin registration

abungay replied:
it does have the whole admin registration

Hajar1 commented on src/test/java/amazin/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,146 @@
+    UserDetailsService userDetailsService;
+
+    @Test
+    public void getLoginTest() throws Exception {
```
Add test goal comment at the top of your test functions; this goes for the whole file

zaidoon1 replied:
this is fine for now

Hajar1 commented on src/test/java/amazin/service/BookServiceTest.java:
```diff
@@ -51,6 +51,7 @@ public void saveTest() {
         verify(repository).save(book);
     }
 
+    @Test
```
@zaidoon1 
This should have been caught when createTest was created; though 👍  @abungay for finding it

Hajar1 reviewed (changes requested):
Good overall. 
Though commenting is lacking all over the place. Please have a look at my changes remarks aside from that excited for user integration 🥇

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+    @Column(name = "user_id", updatable = false, nullable = false)
+    private Long id;
+    @NotNull
+    @Size(min = 1, message = "First name cannot be empty")
```
Should we not constrain max name size?

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+    @Column(unique = true, name="user_email")
+    private String email;
+    @NotNull
+    @Size(min = 8, message = "Password must be at least 8 characters")
```
you could use @min() and @max() annotations

MohamedZalat replied:
This is just a syntax change not a semantic change. Aaron needed to use one and just stay consistent, and he was consistent

Hajar1 commented on src/main/java/amazin/model/User.java:
```diff
@@ -0,0 +1,102 @@
+    @ManyToMany
+    private Set<Role> roles;
+
+    public Long getId() {
```
This is a good place to use the library some of our colleagues were talking about concerning the automatic getters and setters

abungay replied:
not necessary for now

MohamedZalat replied:
This can be a new issue.

Hajar1 commented:
An error message should be in bold red

![Screen Shot 2019-03-29 at 8 29 42 PM](https://user-images.githubusercontent.com/13091952/55268576-6c12fd80-5261-11e9-8e46-f596809a9e89.png)

Hajar1 commented:
The name is not actually reflected in UI after logging in

@abungay  @LauraMachado21 What is your opinions about removing the user profile picture. What is the value of having it?

![Screen Shot 2019-03-29 at 8 31 21 PM](https://user-images.githubusercontent.com/13091952/55268594-a2e91380-5261-11e9-8959-80ecb837f528.png)

abungay commented:
> The name is not actually reflected in UI after logging in
> 
> @abungay @LauraMachado21 What is your opinions about removing the user profile picture. What is the value of having it?
> 
> ![Screen Shot 2019-03-29 at 8 31 21 PM](https://user-images.githubusercontent.com/13091952/55268594-a2e91380-5261-11e9-8959-80ecb837f528.png)

This will be fixed in another PR for issue #63

adhajar committed:
Merge branch 'master' into feat/add-user-auth

adhajar committed:
Merge branch 'master' into feat/add-user-auth
